### PR TITLE
Remove the `PruneUnreferencedOutputs` plan optimizer

### PR DESCRIPTION
**Pull request by kasiafi**
This change removes the legacy PlanOptimizer
`PruneUnreferencedOutputs`, along with the corresponding toggle.

`PruneUnreferencedOutputs` has become redundant due to
set of iterative column-pruning rules. Also, the legacy
optimizer had not been updated for the recently introduced
`PatternRecognitionNode`.

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/analyzer/FeaturesConfig.java [hunk 1]
```diff
@@ -1003,18 +1002,6 @@ public FeaturesConfig setIgnoreDownstreamPreferences(boolean ignoreDownstreamPre
         return this;
     }
 
-    public boolean isIterativeRuleBasedColumnPruning()
```
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/FeaturesConfig.java:1006]: Do we need to add this to list of defunct configs? I'm not sure what's the policy for other such optimizer related configs.

**Review by phd3 — APPROVED** [review]
I only looked at a few tests, and it seems that redundant projections are added. May be the previous optimizer was doing that as a side effect. would adding `RemoveRedundantIdentityProjections` to the test be sufficient? Otherwise we can add a comment in commit-description about the test plan changes.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/FeaturesConfig.java (hunk 1 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/FeaturesConfig.java:994]: > it seems that redundant projections are added

Some of the added projections are pruning, but indeed, some are identity. Thanks for pointing that out.

After investigating the new projections, it seems to me that they all result from enhanced capability of the rule set as compared to the legacy optimizer to prune the `WindowNode`'s source:
- whenever a pruning projection appears under `WindowNode`, it means that the pruning couldn't be pushed down any more
- whenever an identity projection appears under `WindowNode`, it means that a rule inserted a pruning projection, which was further absorbed by the upstream plan (typically, by a `TableScanNode`), making the projection effectively identity. Please note that neither the rule set, nor the legacy optimizer remove identity projections. Only in the legacy scenario the projections didn't appear at all.

To clear the test cases from the identity projections, I followed your suggestion, and made sure that the pruning rules run along with `RemoveRedundantIdentityProjections`. 

The remaining new projections (pruning) had to be added to the expected plans though. I added an explanatory comment in the commit message.

**Review by phd3 — APPROVED** [review]
w.r.t. https://github.com/trinodb/trino/pull/9564#discussion_r744666121, I saw a precedent https://github.com/trinodb/trino/pull/9550 where we didn't add this a defunct config. seems okay to merge this as is.

**Comment by kasiafi** [comment]
@phd3 eventually I added the removed config to the defunct config list.

**Comment by kasiafi** [comment]
@phd3 if you don't object, I'll merge it.

**Review by phd3 — APPROVED** [review]
thanks, lgtm

**Pull request merged**